File: projects/dejavu.ts/dejavu-engine/src/legacy/filter.ts
/**
 * DejaVu 模板引擎过滤器系统
 */

/**
 * 过滤器函数类型
 * @param value 输入值
 * @param args 过滤器参数
 * @returns 过滤后的值
 */
export type Filter = (value: any, args: any[]) => any;

/**
 * 过滤器注册表类
 * 用于管理和应用模板过滤器
 */
export class FilterRegistry {
    /**
     * 已注册的过滤器映射
     */
    private filters: Map<string, Filter>;

    /**
     * 创建过滤器注册表实例
     */
    constructor() {
        this.filters = new Map<string, Filter>();
        this.registerBuiltinFilters();
    }

    /**
     * 注册过滤器
     * @param name 过滤器名称
     * @param filter 过滤器函数
     */
    register(name: string, filter: Filter): void {
        this.filters.set(name, filter);
    }

    /**
     * 获取过滤器
     * @param name 过滤器名称
     * @returns 过滤器函数，如果不存在则返回 undefined
     */
    get(name: string): Filter | undefined {
        return this.filters.get(name);
    }

    /**
     * 检查过滤器是否存在
     * @param name 过滤器名称
     * @returns 是否存在
     */
    has(name: string): boolean {
        return this.filters.has(name);
    }

    /**
     * 应用过滤器
     * @param name 过滤器名称
     * @param value 输入值
     * @param args 过滤器参数
     * @returns 过滤后的值
     */
    apply(name: string, value: any, args: any[]): any {
        const filter = this.filters.get(name);
        if (!filter) {
            throw new Error(`过滤器未找到: ${name}`);
        }
        return filter(value, args);
    }

    /**
     * 注册内置过滤器
     */
    private registerBuiltinFilters(): void {
        this.register("uppercase", uppercaseFilter);
        this.register("lowercase", lowercaseFilter);
        this.register("trim", trimFilter);
        this.register("default", defaultFilter);
        this.register("length", lengthFilter);
        this.register("join", joinFilter);
        this.register("split", splitFilter);
        this.register("replace", replaceFilter);
        this.register("first", firstFilter);
        this.register("last", lastFilter);
        this.register("sort", sortFilter);
        this.register("reverse", reverseFilter);
        this.register("abs", absFilter);
        this.register("round", roundFilter);
        this.register("floor", floorFilter);
        this.register("ceil", ceilFilter);
        this.register("capitalize", capitalizeFilter);
        this.register("title", titleFilter);
        this.register("striptags", striptagsFilter);
        this.register("format", formatFilter);
        this.register("bool", boolFilter);
        this.register("not", notFilter);
    }
}

/**
 * 大写过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 大写字符串
 */
function uppercaseFilter(value: any, args: any[]): any {
    if (typeof value === "string") {
        return value.toUpperCase();
    }
    return value;
}

/**
 * 小写过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 小写字符串
 */
function lowercaseFilter(value: any, args: any[]): any {
    if (typeof value === "string") {
        return value.toLowerCase();
    }
    return value;
}

/**
 * 去除首尾空白过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 去除首尾空白后的字符串
 */
function trimFilter(value: any, args: any[]): any {
    if (typeof value === "string") {
        return value.trim();
    }
    return value;
}

/**
 * 默认值过滤器
 * @param value 输入值
 * @param args 参数，第一个参数为默认值
 * @returns 如果值为假则返回默认值
 */
function defaultFilter(value: any, args: any[]): any {
    if (isTruthy(value)) {
        return value;
    }
    return args.length > 0 ? args[0] : "";
}

/**
 * 长度过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 长度
 */
function lengthFilter(value: any, args: any[]): any {
    if (typeof value === "string") {
        return value.length;
    }
    if (Array.isArray(value)) {
        return value.length;
    }
    if (value && typeof value === "object") {
        return Object.keys(value).length;
    }
    return 0;
}

/**
 * 连接过滤器
 * @param value 输入值
 * @param args 参数，第一个参数为分隔符
 * @returns 连接后的字符串
 */
function joinFilter(value: any, args: any[]): any {
    const separator = args.length > 0 ? String(args[0]) : "";
    if (Array.isArray(value)) {
        return value.map((item) => String(item)).join(separator);
    }
    return "";
}

/**
 * 分割过滤器
 * @param value 输入值
 * @param args 参数，第一个参数为分隔符
 * @returns 分割后的数组
 */
function splitFilter(value: any, args: any[]): any {
    if (typeof value === "string") {
        const separator = args.length > 0 ? String(args[0]) : "";
        return value.split(separator);
    }
    return value;
}

/**
 * 替换过滤器
 * @param value 输入值
 * @param args 参数，第一个为搜索字符串，第二个为替换字符串
 * @returns 替换后的字符串
 */
function replaceFilter(value: any, args: any[]): any {
    if (typeof value === "string") {
        if (args.length >= 2) {
            const search = String(args[0]);
            const replacement = String(args[1]);
            return value.split(search).join(replacement);
        }
        return value;
    }
    return value;
}

/**
 * 第一个元素过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 数组的第一个元素
 */
function firstFilter(value: any, args: any[]): any {
    if (Array.isArray(value)) {
        return value.length > 0 ? value[0] : null;
    }
    return value;
}

/**
 * 最后一个元素过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 数组的最后一个元素
 */
function lastFilter(value: any, args: any[]): any {
    if (Array.isArray(value)) {
        return value.length > 0 ? value[value.length - 1] : null;
    }
    return value;
}

/**
 * 排序过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 排序后的数组
 */
function sortFilter(value: any, args: any[]): any {
    if (Array.isArray(value)) {
        return [...value].sort((a, b) => {
            const aStr = String(a);
            const bStr = String(b);
            if (aStr < bStr) return -1;
            if (aStr > bStr) return 1;
            return 0;
        });
    }
    return value;
}

/**
 * 反转过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 反转后的数组或字符串
 */
function reverseFilter(value: any, args: any[]): any {
    if (Array.isArray(value)) {
        return [...value].reverse();
    }
    if (typeof value === "string") {
        return value.split("").reverse().join("");
    }
    return value;
}

/**
 * 绝对值过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 绝对值
 */
function absFilter(value: any, args: any[]): any {
    if (typeof value === "number") {
        return Math.abs(value);
    }
    return value;
}

/**
 * 四舍五入过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 四舍五入后的整数
 */
function roundFilter(value: any, args: any[]): any {
    if (typeof value === "number") {
        return Math.round(value);
    }
    return value;
}

/**
 * 向下取整过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 向下取整后的整数
 */
function floorFilter(value: any, args: any[]): any {
    if (typeof value === "number") {
        return Math.floor(value);
    }
    return value;
}

/**
 * 向上取整过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 向上取整后的整数
 */
function ceilFilter(value: any, args: any[]): any {
    if (typeof value === "number") {
        return Math.ceil(value);
    }
    return value;
}

/**
 * 首字母大写过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 首字母大写的字符串
 */
function capitalizeFilter(value: any, args: any[]): any {
    if (typeof value === "string" && value.length > 0) {
        return value.charAt(0).toUpperCase() + value.slice(1);
    }
    return value;
}

/**
 * 标题格式过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 每个单词首字母大写的字符串
 */
function titleFilter(value: any, args: any[]): any {
    if (typeof value === "string") {
        return value
            .split(/\s+/)
            .map((word) => {
                if (word.length > 0) {
                    return word.charAt(0).toUpperCase() + word.slice(1);
                }
                return word;
            })
            .join(" ");
    }
    return value;
}

/**
 * 去除 HTML 标签过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 去除 HTML 标签后的字符串
 */
function striptagsFilter(value: any, args: any[]): any {
    if (typeof value === "string") {
        let result = "";
        let inTag = false;
        for (const char of value) {
            if (char === "<") {
                inTag = true;
            } else if (char === ">") {
                inTag = false;
            } else if (!inTag) {
                result += char;
            }
        }
        return result;
    }
    return value;
}

/**
 * 格式化过滤器
 * @param value 输入值（格式字符串）
 * @param args 参数列表
 * @returns 格式化后的字符串
 */
function formatFilter(value: any, args: any[]): any {
    if (typeof value === "string") {
        let result = value;
        for (let i = 0; i < args.length; i++) {
            const placeholder = `{${i}}`;
            result = result.split(placeholder).join(String(args[i]));
        }
        return result;
    }
    return value;
}

/**
 * 布尔值过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 布尔值
 */
function boolFilter(value: any, args: any[]): any {
    return isTruthy(value);
}

/**
 * 逻辑非过滤器
 * @param value 输入值
 * @param args 参数（未使用）
 * @returns 逻辑非结果
 */
function notFilter(value: any, args: any[]): any {
    return !isTruthy(value);
}

/**
 * 判断值是否为真值
 * @param value 输入值
 * @returns 是否为真值
 */
function isTruthy(value: any): boolean {
    if (value === null || value === undefined) {
        return false;
    }
    if (typeof value === "boolean") {
        return value;
    }
    if (typeof value === "number") {
        return value !== 0;
    }
    if (typeof value === "string") {
        return value.length > 0;
    }
    if (Array.isArray(value)) {
        return value.length > 0;
    }
    if (typeof value === "object") {
        return Object.keys(value).length > 0;
    }
    return true;
}

/**
 * 默认过滤器注册表实例
 */
export const filterRegistry = new FilterRegistry();

File: projects/dejavu.rs/dejavu-types/src/values/filter.rs
//! Filter system for DejaVu templates

use crate::{
    errors::{DejavuError, DejavuResult, RuntimeError},
    values::{Context, Value},
};
use std::{collections::HashMap, sync::Arc};

/// Filter trait for template filters
pub trait Filter: Send + Sync + std::fmt::Debug {
    /// Apply the filter to a value
    ///
    /// # Arguments
    /// * `value` - The value to filter
    /// * `args` - Filter arguments
    /// * `context` - Template context
    ///
    /// # Returns
    /// The filtered value
    fn apply(&self, value: Value, args: &[Value], context: &Context) -> DejavuResult<Value>;
}

/// Filter registry for managing filters
#[derive(Debug, Clone)]
pub struct FilterRegistry {
    filters: HashMap<String, Arc<dyn Filter>>,
}

impl FilterRegistry {
    /// Create a new filter registry
    pub fn new() -> Self {
        let mut registry = Self { filters: HashMap::new() };
        registry.register_builtin_filters();
        registry
    }

    /// Register a filter
    pub fn register(&mut self, name: &str, filter: Arc<dyn Filter>) {
        self.filters.insert(name.to_string(), filter);
    }

    /// Get a filter by name
    pub fn get(&self, name: &str) -> Option<&Arc<dyn Filter>> {
        self.filters.get(name)
    }

    /// Apply a filter to a value
    pub fn apply_filter(&self, name: &str, value: Value, args: &[Value], context: &Context) -> DejavuResult<Value> {
        match self.get(name) {
            Some(filter) => filter.apply(value, args, context),
            None => Err(DejavuError::RuntimeError(RuntimeError::FilterNotFound(name.to_string()))),
        }
    }
}

impl Default for FilterRegistry {
    fn default() -> Self {
        Self::new()
    }
}

// Built-in filters

/// Uppercase filter
#[derive(Debug)]
pub struct UppercaseFilter;

impl Filter for UppercaseFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::String(s) => Ok(Value::String(s.to_uppercase())),
            _ => Ok(value),
        }
    }
}

/// Lowercase filter
#[derive(Debug)]
pub struct LowercaseFilter;

impl Filter for LowercaseFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::String(s) => Ok(Value::String(s.to_lowercase())),
            _ => Ok(value),
        }
    }
}

/// Trim filter
#[derive(Debug)]
pub struct TrimFilter;

impl Filter for TrimFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::String(s) => Ok(Value::String(s.trim().to_string())),
            _ => Ok(value),
        }
    }
}

/// Slice filter
#[derive(Debug)]
pub struct SliceFilter;

impl Filter for SliceFilter {
    fn apply(&self, value: Value, args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::String(s) => {
                let start = if args.len() > 0 {
                    match args[0] {
                        Value::Integer(i) => i.max(0) as usize,
                        _ => 0,
                    }
                } else {
                    0
                };

                let end = if args.len() > 1 {
                    match args[1] {
                        Value::Integer(i) => i.max(start as i64) as usize,
                        _ => s.len(),
                    }
                } else {
                    s.len()
                };

                Ok(Value::String(s[start..end.min(s.len())].to_string()))
            }
            _ => Ok(value),
        }
    }
}

/// Round filter
#[derive(Debug)]
pub struct RoundFilter;

impl Filter for RoundFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::Decimal(f) => Ok(Value::Integer(f.round() as i64)),
            Value::Integer(i) => Ok(Value::Integer(i)),
            _ => Ok(value),
        }
    }
}

/// Floor filter
#[derive(Debug)]
pub struct FloorFilter;

impl Filter for FloorFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::Decimal(f) => Ok(Value::Integer(f.floor() as i64)),
            Value::Integer(i) => Ok(Value::Integer(i)),
            _ => Ok(value),
        }
    }
}

/// Ceil filter
#[derive(Debug)]
pub struct CeilFilter;

impl Filter for CeilFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::Decimal(f) => Ok(Value::Integer(f.ceil() as i64)),
            Value::Integer(i) => Ok(Value::Integer(i)),
            _ => Ok(value),
        }
    }
}

/// Bool filter
#[derive(Debug)]
pub struct BoolFilter;

impl Filter for BoolFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        Ok(Value::Bool(value.is_truthy()))
    }
}

/// Default filter
#[derive(Debug)]
pub struct DefaultFilter;

impl Filter for DefaultFilter {
    fn apply(&self, value: Value, args: &[Value], _context: &Context) -> DejavuResult<Value> {
        if value.is_truthy() {
            Ok(value)
        } else {
            if let Some(default_value) = args.get(0) { Ok(default_value.clone()) } else { Ok(Value::String("".to_string())) }
        }
    }
}

/// Length filter
#[derive(Debug)]
pub struct LengthFilter;

impl Filter for LengthFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::String(s) => Ok(Value::Integer(s.len() as i64)),
            Value::Array(arr) => Ok(Value::Integer(arr.len() as i64)),
            Value::Object(obj) => Ok(Value::Integer(obj.len() as i64)),
            _ => Ok(Value::Integer(0)),
        }
    }
}

/// Join filter
#[derive(Debug)]
pub struct JoinFilter;

impl Filter for JoinFilter {
    fn apply(&self, value: Value, args: &[Value], _context: &Context) -> DejavuResult<Value> {
        let separator = args
            .get(0)
            .and_then(|v| match v {
                Value::String(s) => Some(s.as_str()),
                _ => None,
            })
            .unwrap_or("");
        match value {
            Value::Array(arr) => {
                let joined: String = arr
                    .iter()
                    .map(|v| match v {
                        Value::String(s) => s.clone(),
                        _ => "".to_string(),
                    })
                    .collect::<Vec<_>>()
                    .join(separator);
                Ok(Value::String(joined))
            }
            _ => Ok(Value::String("".to_string())),
        }
    }
}

/// Replace filter
#[derive(Debug)]
pub struct ReplaceFilter;

impl Filter for ReplaceFilter {
    fn apply(&self, value: Value, args: &[Value], _context: &Context) -> DejavuResult<Value> {
        if let Value::String(s) = value {
            if args.len() >= 2 {
                let old = match &args[0] {
                    Value::String(s) => s.as_str(),
                    _ => "",
                };
                let new = match &args[1] {
                    Value::String(s) => s.as_str(),
                    _ => "",
                };
                Ok(Value::String(s.replace(old, new)))
            } else {
                Ok(Value::String(s))
            }
        } else {
            Ok(value)
        }
    }
}

/// Split filter
#[derive(Debug)]
pub struct SplitFilter;

impl Filter for SplitFilter {
    fn apply(&self, value: Value, args: &[Value], _context: &Context) -> DejavuResult<Value> {
        if let Value::String(s) = value {
            let separator = args
                .get(0)
                .and_then(|v| match v {
                    Value::String(s) => Some(s.as_str()),
                    _ => None,
                })
                .unwrap_or("")
                .to_string();
            let parts: Vec<Value> = s.split(&separator).map(|part| Value::String(part.to_string())).collect();
            Ok(Value::Array(parts))
        } else {
            Ok(value)
        }
    }
}

/// Capitalize filter
#[derive(Debug)]
pub struct CapitalizeFilter;

impl Filter for CapitalizeFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::String(s) => {
                let mut chars = s.chars();
                match chars.next() {
                    Some(first) => Ok(Value::String(first.to_uppercase().collect::<String>() + chars.as_str())),
                    None => Ok(Value::String(s)),
                }
            }
            _ => Ok(value),
        }
    }
}

/// Title filter
#[derive(Debug)]
pub struct TitleFilter;

impl Filter for TitleFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::String(s) => {
                let title = s
                    .split_whitespace()
                    .map(|word| {
                        let mut chars = word.chars();
                        match chars.next() {
                            Some(first) => first.to_uppercase().collect::<String>() + chars.as_str(),
                            None => word.to_string(),
                        }
                    })
                    .collect::<Vec<_>>()
                    .join(" ");
                Ok(Value::String(title))
            }
            _ => Ok(value),
        }
    }
}

/// Strip tags filter
#[derive(Debug)]
pub struct StripTagsFilter;

impl Filter for StripTagsFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::String(s) => {
                // Simple HTML tag stripping
                let stripped: String = s
                    .chars()
                    .fold((String::new(), false), |(mut result, in_tag), c| match c {
                        '<' => (result, true),
                        '>' => (result, false),
                        _ if in_tag => (result, true),
                        _ => (result + &c.to_string(), false),
                    })
                    .0;
                Ok(Value::String(stripped))
            }
            _ => Ok(value),
        }
    }
}

/// Absolute value filter
#[derive(Debug)]
pub struct AbsFilter;

impl Filter for AbsFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::Integer(i) => Ok(Value::Integer(i.abs())),
            Value::Decimal(f) => Ok(Value::Decimal(f.abs())),
            _ => Ok(value),
        }
    }
}

/// Not filter
#[derive(Debug)]
pub struct NotFilter;

impl Filter for NotFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        Ok(Value::Bool(!value.is_truthy()))
    }
}

/// First filter
#[derive(Debug)]
pub struct FirstFilter;

impl Filter for FirstFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::Array(arr) => Ok(arr.first().cloned().unwrap_or(Value::Null)),
            _ => Ok(value),
        }
    }
}

/// Last filter
#[derive(Debug)]
pub struct LastFilter;

impl Filter for LastFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::Array(arr) => Ok(arr.last().cloned().unwrap_or(Value::Null)),
            _ => Ok(value),
        }
    }
}

/// Sort filter
#[derive(Debug)]
pub struct SortFilter;

impl Filter for SortFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::Array(mut arr) => {
                arr.sort_by(|a, b| {
                    let a_str = match a {
                        Value::String(s) => s.as_str(),
                        _ => "",
                    };
                    let b_str = match b {
                        Value::String(s) => s.as_str(),
                        _ => "",
                    };
                    a_str.cmp(b_str)
                });
                Ok(Value::Array(arr))
            }
            _ => Ok(value),
        }
    }
}

/// Reverse filter
#[derive(Debug)]
pub struct ReverseFilter;

impl Filter for ReverseFilter {
    fn apply(&self, value: Value, _args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::Array(mut arr) => {
                arr.reverse();
                Ok(Value::Array(arr))
            }
            Value::String(s) => Ok(Value::String(s.chars().rev().collect())),
            _ => Ok(value),
        }
    }
}

/// Format filter
#[derive(Debug)]
pub struct FormatFilter;

impl Filter for FormatFilter {
    fn apply(&self, value: Value, args: &[Value], _context: &Context) -> DejavuResult<Value> {
        match value {
            Value::String(format_str) => {
                let mut formatted = format_str;

                // Simple format implementation: {0}, {1}, etc.
                for (i, arg) in args.iter().enumerate() {
                    let placeholder = format!("{{{}}}", i);
                    let arg_str = match arg {
                        Value::String(s) => s.clone(),
                        Value::Integer(i) => i.to_string(),
                        Value::Decimal(f) => f.to_string(),
                        Value::Bool(b) => b.to_string(),
                        _ => "".to_string(),
                    };
                    formatted = formatted.replace(&placeholder, &arg_str);
                }

                Ok(Value::String(formatted))
            }
            _ => Ok(value),
        }
    }
}

/// Implement register_builtin_filters after all filter structs are defined
impl FilterRegistry {
    fn register_builtin_filters(&mut self) {
        // String filters
        self.register("uppercase", Arc::new(UppercaseFilter));
        self.register("lowercase", Arc::new(LowercaseFilter));
        self.register("trim", Arc::new(TrimFilter));
        self.register("slice", Arc::new(SliceFilter));
        self.register("default", Arc::new(DefaultFilter));
        self.register("length", Arc::new(LengthFilter));
        self.register("join", Arc::new(JoinFilter));
        self.register("replace", Arc::new(ReplaceFilter));
        self.register("split", Arc::new(SplitFilter));
        self.register("capitalize", Arc::new(CapitalizeFilter));
        self.register("title", Arc::new(TitleFilter));
        self.register("striptags", Arc::new(StripTagsFilter));
        self.register("format", Arc::new(FormatFilter));

        // Numeric filters
        self.register("round", Arc::new(RoundFilter));
        self.register("floor", Arc::new(FloorFilter));
        self.register("ceil", Arc::new(CeilFilter));
        self.register("abs", Arc::new(AbsFilter));

        // Boolean filters
        self.register("bool", Arc::new(BoolFilter));
        self.register("not", Arc::new(NotFilter));

        // List filters
        self.register("first", Arc::new(FirstFilter));
        self.register("last", Arc::new(LastFilter));
        self.register("sort", Arc::new(SortFilter));
        self.register("reverse", Arc::new(ReverseFilter));
    }
}
